### PR TITLE
Add Alibaba Cloud OpenSearch as a new vector store

### PR DESCRIPTION
**Pull request by zhaoshengbo**
Hello Folks,

Thanks for creating and maintaining this great project. I'm excited to submit this PR to add Alibaba Cloud OpenSearch as a new vector store.

OpenSearch is a one-stop platform to develop intelligent search services. OpenSearch was built based on the large-scale distributed search engine developed by Alibaba. OpenSearch serves more than 500 business cases in Alibaba Group and thousands of Alibaba Cloud customers. OpenSearch helps develop search services in different search scenarios, including e-commerce, O2O, multimedia, the content industry, communities and forums, and big data query in enterprises.

OpenSearch provides the vector search feature. In specific scenarios, especially test question search and image search scenarios, you can use the vector search feature together with the multimodal search feature to improve the accuracy of search results.

This PR includes:

A AlibabaCloudOpenSearch class that can connect to the Alibaba Cloud OpenSearch instance.
add embedings and metadata into a opensearch datasource.
querying by squared euclidean and metadata.
integration tests.
ipython notebook and docs.

I have read your contributing guidelines. And I have passed the tests below

- [x]  make format
- [x]  make lint
- [x]  make coverage
- [x]  make test

### CODE REVIEW

File: langchain/vectorstores/alibabacloud_opensearch.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
+import json
+import logging
+import numbers
+from hashlib import sha1
+from typing import Any, Dict, Iterable, List, Optional, Tuple
+
+from alibabacloud_ha3engine import client, models
```
Comment (hwchase17) [inline langchain/vectorstores/alibabacloud_opensearch.py:7]: these will need to be optional imports

File: pyproject.toml [hunk 2]
```diff
@@ -107,6 +107,7 @@ tigrisdb = {version = "^1.0.0b6", optional = true}
 nebula3-python = {version = "^3.4.0", optional = true}
 langchainplus-sdk = ">=0.0.9"
 awadb = {version = "^0.3.3", optional = true}
+alibabacloud-ha3engine = "^1.3.4"
```
Comment (hwchase17) [inline pyproject.toml:110]: this will need to be optional

File: langchain/vectorstores/alibabacloud_opensearch.py (hunk 1 above)
Comment (zhaoshengbo) [reply langchain/vectorstores/alibabacloud_opensearch.py]: thanks for reviewing, I have fixed it.

File: pyproject.toml (hunk 2 above)
Comment (zhaoshengbo) [reply pyproject.toml]: thanks for reviewing, I have fixed it.

**Comment by rlancemartin** [comment]
Thanks for the contribution! I will review and help get this merged tomorrow.

**Comment by rlancemartin** [comment]
@zhaoshengbo can you please rebase on current master? 

Your branch has some merge conflicts, as you can see in the tests.

I checked out your branched and attempted to rebase, confirmed these:
```
CONFLICT (file location): docs/integrations/alibabacloud_opensearch.md added in f963630f (add alibaba cloud openseearch vector store) inside a directory that was renamed in HEAD, suggesting it should perhaps be moved to docs/extras/ecosystem/integrations/alibabacloud_opensearch.md.
CONFLICT (file location): docs/modules/indexes/vectorstores/examples/alibabacloud_opensearch.ipynb added in f963630f (add alibaba cloud openseearch vector store) inside a directory that was renamed in HEAD, suggesting it should perhaps be moved to docs/extras/modules/data_connection/vectorstores/integrations/alibabacloud_opensearch.ipynb.
Auto-merging langchain/vectorstores/__init__.py
CONFLICT (content): Merge conflict in langchain/vectorstores/__init__.py
Auto-merging poetry.lock
CONFLICT (content): Merge conflict in poetry.lock
Auto-merging pyproject.toml
CONFLICT (content): Merge conflict in pyproject.toml
```

**Comment by rlancemartin** [comment]
Just rebased for you, and kicked off testing.

**Review by rlancemartin — APPROVED** [review]
Nice work!

**Comment by zhaoshengbo** [comment]
> Nice work!

thanks for your help

**Comment by zhaoshengbo** [comment]
> Just rebased for you, and kicked off testing. Hopefully this clears us to merge.

thanks for your help

**Pull request merged**